File: frontend/src/types/hr.ts
// 前端 API 型別定義擴展
// HR, Audit, Facility, Calendar 相關型別

// ============================================
// Audit Types
// ============================================

export interface UserActivityLog {
    id: string;
    actor_user_id: string | null;
    actor_email: string | null;
    actor_display_name: string | null;
    event_category: string;
    event_type: string;
    event_severity: string;
    entity_type: string | null;
    entity_id: string | null;
    entity_display_name: string | null;
    before_data: Record<string, unknown> | null;
    after_data: Record<string, unknown> | null;
    ip_address: string | null;
    user_agent: string | null;
    request_path: string | null;
    is_suspicious: boolean;
    suspicious_reason: string | null;
    created_at: string;
}

export interface LoginEventWithUser {
    id: string;
    user_id: string | null;
    email: string;
    user_name: string | null;
    event_type: string;
    ip_address: string | null;
    user_agent: string | null;
    device_type: string | null;
    browser: string | null;
    os: string | null;
    is_unusual_time: boolean;
    is_unusual_location: boolean;
    is_new_device: boolean;
    failure_reason: string | null;
    created_at: string;
}

export interface SessionWithUser {
    id: string;
    user_id: string;
    user_email: string;
    user_name: string;
    started_at: string;
    ended_at: string | null;
    last_activity_at: string;
    ip_address: string | null;
    user_agent: string | null;
    page_view_count: number;
    action_count: number;
    is_active: boolean;
    ended_reason: string | null;
}

export interface SecurityAlert {
    id: string;
    alert_type: string;
    severity: string;
    title: string;
    description: string | null;
    user_id: string | null;
    status: string;
    resolved_by: string | null;
    resolved_at: string | null;
    resolution_notes: string | null;
    created_at: string;
}

export interface AuditDashboardStats {
    active_users_today: number;
    active_users_week: number;
    active_users_month: number;
    total_logins_today: number;
    failed_logins_today: number;
    active_sessions: number;
    open_alerts: number;
    critical_alerts: number;
}

// ============================================
// HR Types
// ============================================

export interface AttendanceWithUser {
    id: string;
    user_id: string;
    user_email: string;
    user_name: string;
    work_date: string;
    clock_in_time: string | null;
    clock_out_time: string | null;
    regular_hours: number | null;
    overtime_hours: number | null;
    status: string;
    remark: string | null;
    is_corrected: boolean;
}

export interface OvertimeWithUser {
    id: string;
    user_id: string;
    user_email: string;
    user_name: string;
    overtime_date: string;
    start_time: string;
    end_time: string;
    hours: number;
    overtime_type: string;
    multiplier: number;
    comp_time_hours: number;
    comp_time_expires_at: string;
    status: string;
    reason: string;
}

export interface LeaveRequestWithUser {
    id: string;
    user_id: string;
    user_email: string;
    user_name: string;
    proxy_user_id: string | null;
    proxy_user_name: string | null;
    leave_type: string;
    start_date: string;
    end_date: string;
    total_days: number;
    reason: string;
    is_urgent: boolean;
    is_retroactive: boolean;
    status: string;
    current_approver_id: string | null;
    current_approver_name: string | null;
    submitted_at: string | null;
    created_at: string;
}

export interface AnnualLeaveBalanceView {
    entitlement_year: number;
    entitled_days: number;
    used_days: number;
    remaining_days: number;
    expires_at: string;
    days_until_expiry: number;
}

export interface CompTimeBalanceView {
    id: string;
    earned_date: string;
    original_hours: number;
    used_hours: number;
    remaining_hours: number;
    expires_at: string;
    days_until_expiry: number;
}

export interface BalanceSummary {
    user_id: string;
    user_name: string;
    annual_leave_total: number;
    annual_leave_used: number;
    annual_leave_remaining: number;
    comp_time_total: number;
    comp_time_used: number;
    comp_time_remaining: number;
    expiring_soon_days: number;
    expiring_soon_hours: number;
}

// Leave Type 顯示名稱映射
export const LEAVE_TYPE_NAMES: Record<string, string> = {
    ANNUAL: '特休假',
    PERSONAL: '事假',
    SICK: '病假',
    COMPENSATORY: '補休假',
    MARRIAGE: '婚假',
    BEREAVEMENT: '喪假',
    MATERNITY: '產假',
    PATERNITY: '陪產假',
    MENSTRUAL: '生理假',
    OFFICIAL: '公假',
    UNPAID: '無薪假',
};

export const LEAVE_STATUS_NAMES: Record<string, string> = {
    DRAFT: '草稿',
    PENDING_L1: '待一級審核',
    PENDING_L2: '待二級審核',
    PENDING_HR: '待行政審核',
    PENDING_GM: '待總經理核准',
    APPROVED: '已核准',
    REJECTED: '已駁回',
    CANCELLED: '已取消',
    REVOKED: '已銷假',
};

// ============================================
// Facility Types
// ============================================

export interface Species {
    id: string;
    code: string;
    name: string;
    name_en: string | null;
    icon: string | null;
    is_active: boolean;
    config: Record<string, unknown> | null;
    sort_order: number;
}

export interface Facility {
    id: string;
    code: string;
    name: string;
    address: string | null;
    phone: string | null;
    contact_person: string | null;
    is_active: boolean;
}

export interface BuildingWithFacility {
    id: string;
    facility_id: string;
    facility_code: string;
    facility_name: string;
    code: string;
    name: string;
    description: string | null;
    is_active: boolean;
    sort_order: number;
}

export interface ZoneWithBuilding {
    id: string;
    building_id: string;
    building_code: string;
    building_name: string;
    facility_id: string;
    facility_name: string;
    code: string;
    name: string | null;
    color: string | null;
    is_active: boolean;
    sort_order: number;
}

export interface PenDetails {
    id: string;
    code: string;
    name: string | null;
    capacity: number;
    current_count: number;
    status: string;
    zone_id: string;
    zone_code: string;
    zone_name: string | null;
    zone_color: string | null;
    building_id: string;
    building_code: string;
    building_name: string;
    facility_id: string;
    facility_code: string;
    facility_name: string;
}

export interface DepartmentWithManager {
    id: string;
    code: string;
    name: string;
    parent_id: string | null;
    parent_name: string | null;
    manager_id: string | null;
    manager_name: string | null;
    is_active: boolean;
    sort_order: number;
}

// ============================================
// Calendar Types
// ============================================

export interface CalendarSyncStatus {
    is_configured: boolean;
    sync_enabled: boolean;
    calendar_id: string;
    last_sync_at: string | null;
    last_sync_status: string | null;
    next_sync_at: string | null;
    pending_syncs: number;
    pending_conflicts: number;
    recent_errors: number;
}

export interface ConflictWithDetails {
    id: string;
    leave_request_id: string | null;
    user_name: string | null;
    leave_type: string | null;
    conflict_type: string;
    difference_summary: string | null;
    status: string;
    detected_at: string;
}

export interface CalendarSyncHistory {
    id: string;
    job_type: string;
    triggered_by: string | null;
    started_at: string;
    completed_at: string | null;
    duration_ms: number | null;
    status: string;
    events_created: number;
    events_updated: number;
    events_deleted: number;
    conflicts_detected: number;
    errors_count: number;
}

export interface CalendarEvent {
    id: string;
    summary: string;
    start: string;
    end: string;
    all_day: boolean;
    description?: string;
    location?: string;
    color_id?: string;
    html_link?: string;
}

File: backend/src/models/hr.rs
// HR 模組 Models
// 包含：Attendance, Overtime, Leave, Balances

use chrono::{DateTime, NaiveDate, NaiveTime, Utc};
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

// ============================================
// Attendance (出勤)
// ============================================

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct AttendanceRecord {
    pub id: Uuid,
    pub user_id: Uuid,
    pub work_date: NaiveDate,
    pub clock_in_time: Option<DateTime<Utc>>,
    pub clock_out_time: Option<DateTime<Utc>>,
    pub regular_hours: Option<Decimal>,
    pub overtime_hours: Option<Decimal>,
    pub status: String,
    pub clock_in_source: Option<String>,
    pub clock_in_ip: Option<String>,
    pub clock_out_source: Option<String>,
    pub clock_out_ip: Option<String>,
    pub remark: Option<String>,
    pub is_corrected: bool,
    pub corrected_by: Option<Uuid>,
    pub corrected_at: Option<DateTime<Utc>>,
    pub correction_reason: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Serialize, FromRow)]
pub struct AttendanceWithUser {
    pub id: Uuid,
    pub user_id: Uuid,
    pub user_email: String,
    pub user_name: String,
    pub work_date: NaiveDate,
    pub clock_in_time: Option<DateTime<Utc>>,
    pub clock_out_time: Option<DateTime<Utc>>,
    pub regular_hours: Option<Decimal>,
    pub overtime_hours: Option<Decimal>,
    pub status: String,
    pub remark: Option<String>,
    pub is_corrected: bool,
}

#[derive(Debug, Deserialize)]
pub struct AttendanceQuery {
    pub user_id: Option<Uuid>,
    pub from: Option<NaiveDate>,
    pub to: Option<NaiveDate>,
    pub status: Option<String>,
    pub page: Option<i64>,
    pub per_page: Option<i64>,
}

#[derive(Debug, Deserialize)]
pub struct ClockInRequest {
    pub source: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct ClockOutRequest {
    pub source: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct AttendanceCorrectionRequest {
    pub clock_in_time: Option<DateTime<Utc>>,
    pub clock_out_time: Option<DateTime<Utc>>,
    pub reason: String,
}

// ============================================
// Overtime (加班)
// ============================================

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct OvertimeRecord {
    pub id: Uuid,
    pub user_id: Uuid,
    pub attendance_id: Option<Uuid>,
    pub overtime_date: NaiveDate,
    pub start_time: DateTime<Utc>,
    pub end_time: DateTime<Utc>,
    pub hours: Decimal,
    pub overtime_type: String,
    pub multiplier: Decimal,
    pub comp_time_hours: Decimal,
    pub comp_time_expires_at: NaiveDate,
    pub comp_time_used_hours: Decimal,
    pub status: String,
    pub submitted_at: Option<DateTime<Utc>>,
    pub approved_by: Option<Uuid>,
    pub approved_at: Option<DateTime<Utc>>,
    pub rejected_by: Option<Uuid>,
    pub rejected_at: Option<DateTime<Utc>>,
    pub rejection_reason: Option<String>,
    pub reason: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Serialize, FromRow)]
pub struct OvertimeWithUser {
    pub id: Uuid,
    pub user_id: Uuid,
    pub user_email: String,
    pub user_name: String,
    pub overtime_date: NaiveDate,
    pub start_time: DateTime<Utc>,
    pub end_time: DateTime<Utc>,
    pub hours: Decimal,
    pub overtime_type: String,
    pub multiplier: Decimal,
    pub comp_time_hours: Decimal,
    pub comp_time_expires_at: NaiveDate,
    pub status: String,
    pub reason: String,
}

#[derive(Debug, Deserialize)]
pub struct OvertimeQuery {
    pub user_id: Option<Uuid>,
    pub status: Option<String>,
    pub from: Option<NaiveDate>,
    pub to: Option<NaiveDate>,
    pub pending_approval: Option<bool>,
    pub page: Option<i64>,
    pub per_page: Option<i64>,
}

#[derive(Debug, Deserialize)]
pub struct CreateOvertimeRequest {
    pub overtime_date: NaiveDate,
    pub start_time: NaiveTime,
    pub end_time: NaiveTime,
    pub overtime_type: String,
    pub reason: String,
}

#[derive(Debug, Deserialize)]
pub struct UpdateOvertimeRequest {
    pub start_time: Option<NaiveTime>,
    pub end_time: Option<NaiveTime>,
    pub overtime_type: Option<String>,
    pub reason: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct RejectOvertimeRequest {
    pub reason: String,
}

// ============================================
// Leave (請假)
// ============================================

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::Type, PartialEq)]
#[sqlx(type_name = "leave_type", rename_all = "SCREAMING_SNAKE_CASE")]
pub enum LeaveType {
    Annual,
    Personal,
    Sick,
    Compensatory,
    Marriage,
    Bereavement,
    Maternity,
    Paternity,
    Menstrual,
    Official,
    Unpaid,
}

impl LeaveType {
    pub fn display_name(&self) -> &'static str {
        match self {
            LeaveType::Annual => "特休假",
            LeaveType::Personal => "事假",
            LeaveType::Sick => "病假",
            LeaveType::Compensatory => "補休假",
            LeaveType::Marriage => "婚假",
            LeaveType::Bereavement => "喪假",
            LeaveType::Maternity => "產假",
            LeaveType::Paternity => "陪產假",
            LeaveType::Menstrual => "生理假",
            LeaveType::Official => "公假",
            LeaveType::Unpaid => "無薪假",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::Type, PartialEq)]
#[sqlx(type_name = "leave_status", rename_all = "SCREAMING_SNAKE_CASE")]
pub enum LeaveStatus {
    Draft,
    PendingL1,
    PendingL2,
    PendingHr,
    PendingGm,
    Approved,
    Rejected,
    Cancelled,
    Revoked,
}

impl LeaveStatus {
    pub fn display_name(&self) -> &'static str {
        match self {
            LeaveStatus::Draft => "草稿",
            LeaveStatus::PendingL1 => "待一級審核",
            LeaveStatus::PendingL2 => "待二級審核",
            LeaveStatus::PendingHr => "待行政審核",
            LeaveStatus::PendingGm => "待總經理核准",
            LeaveStatus::Approved => "已核准",
            LeaveStatus::Rejected => "已駁回",
            LeaveStatus::Cancelled => "已取消",
            LeaveStatus::Revoked => "已銷假",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct LeaveRequest {
    pub id: Uuid,
    pub user_id: Uuid,
    pub proxy_user_id: Option<Uuid>,  // 代理人
    pub leave_type: String, // 用 String 避免 sqlx enum 問題
    pub start_date: NaiveDate,
    pub end_date: NaiveDate,
    pub start_time: Option<NaiveTime>,
    pub end_time: Option<NaiveTime>,
    pub total_days: Decimal,
    pub total_hours: Option<Decimal>,
    pub reason: String,
    pub supporting_documents: Option<serde_json::Value>,
    pub comp_time_source_ids: Option<Vec<Uuid>>,
    pub annual_leave_source_id: Option<Uuid>,
    pub is_urgent: bool,
    pub is_retroactive: bool,
    pub status: String,
    pub current_approver_id: Option<Uuid>,
    pub submitted_at: Option<DateTime<Utc>>,
    pub approved_at: Option<DateTime<Utc>>,
    pub rejected_at: Option<DateTime<Utc>>,
    pub cancelled_at: Option<DateTime<Utc>>,
    pub revoked_at: Option<DateTime<Utc>>,
    pub cancellation_reason: Option<String>,
    pub revocation_reason: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Serialize, FromRow)]
pub struct LeaveRequestWithUser {
    pub id: Uuid,
    pub user_id: Uuid,
    pub user_email: String,
    pub user_name: String,
    pub proxy_user_id: Option<Uuid>,
    pub proxy_user_name: Option<String>,
    pub leave_type: String,
    pub start_date: NaiveDate,
    pub end_date: NaiveDate,
    pub total_days: Decimal,
    pub reason: String,
    pub is_urgent: bool,
    pub is_retroactive: bool,
    pub status: String,
    pub current_approver_id: Option<Uuid>,
    pub current_approver_name: Option<String>,
    pub submitted_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Deserialize)]
pub struct LeaveQuery {
    pub user_id: Option<Uuid>,
    pub status: Option<String>,
    pub leave_type: Option<String>,
    pub from: Option<NaiveDate>,
    pub to: Option<NaiveDate>,
    pub pending_approval: Option<bool>,
    pub page: Option<i64>,
    pub per_page: Option<i64>,
}

#[derive(Debug, Deserialize)]
pub struct CreateLeaveRequest {
    pub leave_type: String,
    pub start_date: NaiveDate,
    pub end_date: NaiveDate,
    pub start_time: Option<NaiveTime>,
    pub end_time: Option<NaiveTime>,
    pub total_days: f64,
    pub total_hours: Option<f64>,
    pub reason: Option<String>,  // 特休假不用填理由
    pub supporting_documents: Option<Vec<String>>,  // 附件圖片 URLs
    pub is_urgent: Option<bool>,
    pub is_retroactive: Option<bool>,
    pub proxy_user_id: Option<Uuid>,  // 代理人
}

#[derive(Debug, Deserialize)]
pub struct UpdateLeaveRequest {
    pub start_date: Option<NaiveDate>,
    pub end_date: Option<NaiveDate>,
    pub start_time: Option<NaiveTime>,
    pub end_time: Option<NaiveTime>,
    pub total_days: Option<f64>,
    pub total_hours: Option<f64>,
    pub reason: Option<String>,
    pub proxy_user_id: Option<Uuid>,  // 代理人
}

#[derive(Debug, Deserialize)]
pub struct ApproveLeaveRequest {
    pub comments: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct RejectLeaveRequest {
    pub reason: String,
}

#[derive(Debug, Deserialize)]
pub struct CancelLeaveRequest {
    pub reason: Option<String>,
}

// ============================================
// Leave Approvals (審核記錄)
// ============================================

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct LeaveApproval {
    pub id: Uuid,
    pub leave_request_id: Uuid,
    pub approver_id: Uuid,
    pub approval_level: String,
    pub action: String,
    pub comments: Option<String>,
    pub created_at: DateTime<Utc>,
}

// ============================================
// Balances (餘額)
// ============================================

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct AnnualLeaveEntitlement {
    pub id: Uuid,
    pub user_id: Uuid,
    pub entitlement_year: i32,
    pub entitled_days: Decimal,
    pub used_days: Decimal,
    pub expires_at: NaiveDate,
    pub calculation_basis: Option<String>,
    pub seniority_years: Option<Decimal>,
    pub is_expired: bool,
    pub expired_days: Decimal,
    pub expiry_processed_at: Option<DateTime<Utc>>,
    pub notes: Option<String>,
    pub created_by: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct CompTimeBalance {
    pub id: Uuid,
    pub user_id: Uuid,
    pub overtime_record_id: Uuid,
    pub original_hours: Decimal,
    pub used_hours: Decimal,
    pub earned_date: NaiveDate,
    pub expires_at: NaiveDate,
    pub is_expired: bool,
    pub expired_hours: Decimal,
    pub converted_to_pay: bool,
    pub expiry_processed_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Serialize)]
pub struct AnnualLeaveBalanceView {
    pub entitlement_year: i32,
    pub entitled_days: f64,
    pub used_days: f64,
    pub remaining_days: f64,
    pub expires_at: NaiveDate,
    pub days_until_expiry: i32,
    pub is_expired: bool,  // 是否已過期（待補償）
}

#[derive(Debug, Serialize)]
pub struct CompTimeBalanceView {
    pub id: Uuid,
    pub earned_date: NaiveDate,
    pub original_hours: f64,
    pub used_hours: f64,
    pub remaining_hours: f64,
    pub expires_at: NaiveDate,
    pub days_until_expiry: i32,
}

#[derive(Debug, Serialize)]
pub struct BalanceSummary {
    pub user_id: Uuid,
    pub user_name: String,
    pub annual_leave_total: f64,
    pub annual_leave_used: f64,
    pub annual_leave_remaining: f64,
    pub comp_time_total: f64,
    pub comp_time_used: f64,
    pub comp_time_remaining: f64,
    pub expiring_soon_days: f64,
    pub expiring_soon_hours: f64,
}

#[derive(Debug, Deserialize)]
pub struct BalanceQuery {
    pub user_id: Option<Uuid>,
    pub year: Option<i32>,
    pub include_expired: Option<bool>,
}

#[derive(Debug, Deserialize)]
pub struct CreateAnnualLeaveRequest {
    pub user_id: Uuid,
    pub entitlement_year: i32,
    pub entitled_days: f64,
    pub hire_date: Option<NaiveDate>,  // 到職日，用於計算到期日（到職週年日 + 2年）
    pub calculation_basis: Option<String>,
    pub notes: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct AdjustBalanceRequest {
    pub adjustment_days: f64,
    pub reason: String,
}

/// 過期特休假報表（待補償）
#[derive(Debug, Serialize)]
pub struct ExpiredLeaveReport {
    pub user_id: Uuid,
    pub user_name: String,
    pub user_email: String,
    pub entitlement_year: i32,
    pub entitled_days: f64,
    pub used_days: f64,
    pub remaining_days: f64,  // 待補償天數
    pub expires_at: NaiveDate,
}

// ============================================
// Dashboard Calendar (儀表板日曆)
// ============================================

#[derive(Debug, Clone, Serialize)]
pub struct TodayLeaveInfo {
    pub user_id: Uuid,
    pub user_name: String,
    pub leave_type: String,
    pub leave_type_display: String,
    pub is_all_day: bool,
    pub start_date: NaiveDate,
    pub end_date: NaiveDate,
}

#[derive(Debug, Serialize)]
pub struct DashboardCalendarData {
    pub today: NaiveDate,
    pub today_leaves: Vec<TodayLeaveInfo>,
    pub today_events: Vec<crate::models::calendar::CalendarEvent>,
    pub upcoming_leaves: Vec<TodayLeaveInfo>,
}
